src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,505 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/BasicConfig.h>
+#include <ripple/basics/impl/PerfLogImp.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/json/json_writer.h>
+#include <ripple/json/to_string.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <atomic>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+namespace ripple {
+namespace perf {
+
+void
+PerfLogImp::Counters::populateRpc(std::vector<char const*> const& labels)
+{
+    for ( auto label : labels)
+    {
+        auto const &elem = rpc_.emplace(label, Rpc(label));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+void
+PerfLogImp::Counters::populateJq(
+    std::unordered_map<std::underlying_type_t<JobType>,
+    std::string> const& jobs)
+{
+    for (auto const& job : jobs)
+    {
+        auto elem = jq_.emplace(job.first, Jq(job.second));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+Json::Value
+PerfLogImp::Counters::countersJson() const
+{
+    Json::Value rpcobj(Json::objectValue);
+    // totalRpc represents all rpc methods. All that started, finished, etc.
+    Rpc totalRpc("total");
+    for (auto const& proc : rpc_)
+    {
+        Json::Value p(Json::objectValue);
+        p[jss::started] = std::to_string(proc.second.started);
```

Comment:
Thanks for your note about the thread safety of members `rpc_` and `jq_`.  Unfortunately I think you've got a data race here, at least as far as the C++17 Standard is concerned.  All of the writes to values in `rpc_` and `jq_` are protected by their appropriate `mut`.  But these _reads_ are not guarded.  According to the standard (N4660), section 4.7.1 paragraph 2, "Expression evaluations conflict if one of them modifies a memory location (4.4) and the other one reads or modifies the same memory location."  So unguarded reads can produce data races, which in turn produce undefined behavior.

Probably the simplest fix for this race condition is to guard these reads with the appropriate `mut`.  That makes me squirm a bit, since you'll be doing the `toString()` calls and JSON production under the lock.  I think it's okay to call those functions while holding a lock, it just means you'll be holding the lock for a while (which will any writing threads).  String processing and heap allocation are slow.

Another approach might be to rethink your `rpc_` and `jq_` containers so they are cheap to copy (and move the mutexes out).  Then you could use the same approach to the reads as you used here: https://github.com/ripple/rippled/pull/2383/files#diff-458444c50ff4bc52cb805f402d021b9bR138.

Or you may have a third approach in mind.  Your call.

Reply:
The only unguarded reads are to the key then the corresponding struct (Rpc or Jq). All of those are created before multiple threads are spawned. Therefore, they won't be subject to modification. The only things that are modified after multiple workers are created are the uint64_t data members. Those are protected by their corresponding mutexes. The location in memory for each uint64_t is going to stay consistent through each's lifetime, as members of the struct to which they belong.

Reply:
Oh, good catch...sorry about the previous comment, yes these are ungarded reads.

Reply:
@scottschurr Good catch on guarding the reads: fixed
scons vcxproj: fixed
commit message: fixed
JobTypes singleton: scons vcxproj by cherry-pick